pkg/river/internal/value/decode.go

```diff
@@ -13,6 +13,24 @@ import (
 
 // The Defaulter interface allows a type to implement default functionality
 // in River evaluation.
+//
+// SetToDefault() will be called only for block and body river types. To
+// set arguments to their defaults, use the SetToDefault() method the arguments'
+// parent block.
+//
+// Care should be taken when using nested optional blocks with defaults:
+//   - when an inner block is provided with empty body, the inner type's
+//     defaults will be used.
+//   - when an inner block is not provided at all, the outer type's defaults
+//     will be used.
+//   - if the defaults are not defined, zero values (for structs) or
+//     nil (for pointers) will be set for fields that are not present.
+//     This means that, for example, if an outer type doesn't define defaults,
+//     but the inner type does, the inner type's defaults will not be used when
+//     the block is not provided at all.
+//
+// To avoid confusion, it is strongly recommended to ensure that the
+// SetToDefault method sets the same defaults for the outer and the inner types. .
```

Comment:
I think this is a bit long. What do you think about this?

```suggestion
// Defaulter is only used for block and body River types.
//
// If a River block or body wraps a type implementing Defaulter, the 
// wrapping type must also implement Defaulter to propagate the default 
// value of of the wrapped type, otherwise defaults for the wrapped type 
// only apply when the wrapped type is defined in the River config file. 
```

Reply:
I personally find it harder to understand it in this compressed form. Also I would like to clearly state that if we have e.g. `inner {}` in River then the inner type's defaults will be used, but when we have nothing `""` then the outer type's defaults will be used.

> Defaulter is only used for block and body River types.
You can use Defaulter anywhere, but it won't be invoked unless it's on a block or body. But also, the defaulter can be used to set the arguments inside the block and body - so in a way it can be used 'with' arguments, but you shouldn't make your argument types implement it. 

> the wrapping type must also implement Defaulter to propagate the default
It does get funny with pointers though - I think it's fine if the outer type sets the inner field to `nil`, but it can be added with its defaults (from the inner type) if one provides `inner { }`. So I wouldn't say that the wrapping type must implement Defaulter. If it was a requirement, we probably should check that it holds and error when it's violated?

Reply:
> Also I would like to clearly state that if we have e.g. inner {} in River then the inner type's defaults will be used, but when we have nothing "" then the outer type's defaults will be used.

I tried to say that here: 

> defaults for the wrapped type only apply when the wrapped type is defined in the River config file. 

Is there a clearer way of describing this succinctly?

Reply:
How about something like this?

```
// Defaulter will be called only on block and body river types.
//
// When using nested blocks, the wrapping type should also implement the 
// Defaulter to propagate the defaults of the wrapped type. Otherwise, the 
// wrapping type's defaults will be used when the wrapped field is not defined
// in the River config, or the wrapped type's defaults will be used if the 
// wrapped field is defined in the River config.
```
Heavily inspired by your suggestion, but slightly different.

Reply:
I think we should make this a "must" rather than a "should" w/r/t propagating defaults. 

I'd like to tweak the "Otherwise" part a bit to make it rotate around what happens when the wrapped block is defined compared to when defaults are used:

```
// Defaulter will be called only on block and body river types.
//
// When using nested blocks, the wrapping type must also implement the 
// Defaulter to propagate the defaults of the wrapped type. Otherwise, 
// defaults used for the wrapped type become inconsistent: 
//
// - If the the wrapped block is NOT defined in the River config, the wrapping 
//   type's defaults are used. 
// - If the wrapped block IS defined in the River config, the wrapped type's 
//   defaults are used. 
```

WDYT? I reintroduced the bullet list to hopefully make it easier to scan as the paragraph got longer while still keeping the overall additions to the comment shorter than before.